src/Module/Profile/Photos/Upload.php

```diff
@@ -0,0 +1,277 @@
+			$this->exit(401, $this->l10n->t('Unable to process image.'));
+		}
+
+		$image->orient($src);
+		@unlink($src);
+
+		$max_length = $this->config->get('system', 'max_image_length');
+		if ($max_length > 0) {
+			$image->scaleDown($max_length);
+			$filesize = strlen($image->asString());
+			$this->logger->info('File upload: Scaling picture to new size', ['max_length' => $max_length]);
+		}
+
+		$width  = $image->getWidth();
+		$height = $image->getHeight();
+
+		$maximagesize = $this->config->get('system', 'maximagesize');
+
+		if (!empty($maximagesize) && $filesize > $maximagesize) {
+			// Scale down to multiples of 640 until the maximum size isn't exceeded anymore
+			foreach ([5120, 2560, 1280, 640] as $pixels) {
+				if ($filesize > $maximagesize && max($width, $height) > $pixels) {
+					$this->logger->info('Resize', ['size' => $filesize, 'width' => $width, 'height' => $height, 'max' => $maximagesize, 'pixels' => $pixels]);
+					$image->scaleDown($pixels);
+					$filesize = strlen($image->asString());
+					$width    = $image->getWidth();
+					$height   = $image->getHeight();
+				}
+			}
+
+			if ($filesize > $maximagesize) {
+				@unlink($src);
+				$this->logger->notice('Image size is too big', ['size' => $filesize, 'max' => $maximagesize]);
+				$this->exit(401, $this->l10n->t('Image exceeds size limit of %s', Strings::formatBytes($maximagesize)));
+			}
+		}
+
+		$resource_id = Photo::newResource();
+
+		$smallest = 0;
+
+		// If we don't have an album name use the Wall Photos album
+		if (!strlen($album)) {
+			$album = $this->l10n->t('Wall Photos');
+		}
+
+		$defperm = '<' . $default_cid . '>';
+
+		$result = Photo::store($image, $page_owner_uid, $visitor, $resource_id, $filename, $album, 0, Photo::DEFAULT, $defperm);
+		if (!$result) {
+			$this->logger->warning('Photo::store() failed', ['result' => $result]);
+			$this->exit(401, $this->l10n->t('Image upload failed.'));
+		}
+
+		if ($width > 640 || $height > 640) {
+			$image->scaleDown(640);
+			$result = Photo::store($image, $page_owner_uid, $visitor, $resource_id, $filename, $album, 1, Photo::DEFAULT, $defperm);
+			if ($result) {
+				$smallest = 1;
+			}
+		}
+
+		if ($width > 320 || $height > 320) {
+			$image->scaleDown(320);
+			$result = Photo::store($image, $page_owner_uid, $visitor, $resource_id, $filename, $album, 2, Photo::DEFAULT, $defperm);
+			if ($result && ($smallest == 0)) {
+				$smallest = 2;
+			}
+		}
+
+		$this->logger->info('upload done');
+		$this->exit(200, "\n\n" . '[url=' . $this->baseUrl . '/photos/' . $user['nickname'] . '/image/' . $resource_id . '][img]' . $this->baseUrl . "/photo/$resource_id-$smallest." . $image->getExt() . "[/img][/url]\n\n");
+	}
+
+	/**
+	 * @param int    $httpCode
+	 * @param string $message
+	 * @param bool   $systemMessage
+	 * @return void
+	 * @throws InternalServerErrorException
+	 */
+	private function exit(int $httpCode, string $message, bool $systemMessage = false): void
```

Comment:
Can you check if `$this->response` is useful here?

As discussed yesterday, it's currently not fully useful, but in the end, everything should be a Response and no module or class should directly exit the process -> testability:-)

Reply:
After reviewing what happens with the embedded `$this->response` in `BaseModule->run()`, I ended up simplifying the exit method and removing the dependency to the `Core\System` class. 👍